tokio/src/sync/watch.rs

```diff
@@ -1302,8 +1311,29 @@ impl<T> Sender<T> {
 
 impl<T> Drop for Sender<T> {
     fn drop(&mut self) {
-        self.shared.state.set_closed();
-        self.shared.notify_rx.notify_waiters();
+        if self.shared.ref_count_tx.fetch_sub(1, AcqRel) == 1 {
+            // Now that the sender's ref count is zero, so we attempt to close the sender.
+
+            // Fetch the latest state.
+            let current_state = self.shared.state.load();
+            let currrent_state_inner = current_state.0;
+
+            if self
+                .shared
+                .state
+                .0
+                .compare_exchange(
+                    currrent_state_inner,
+                    currrent_state_inner | CLOSED_BIT,
+                    AcqRel,
+                    Relaxed,
+                )
+                .is_ok()
+            {
+                // Only when last sender is dropped, we notify the receivers.
+                self.shared.notify_rx.notify_waiters();
+            }
+        }
```

Comment:
Hmm, maybe I'm mistaken here. I'll think more about it.

Reply:
Why does it not do this?
```Rust
if self.shared.ref_count_tx.fetch_sub(1, AcqRel) == 1 {
        self.shared.state.set_closed();
        self.shared.notify_rx.notify_waiters();
}
```
That's what it did before, so why change it?

Reply:
Ah, thanks. By allowing sender clone, I thought I would also need CAS when we set close bit, but now I think the if condition will be true only when all other senders have been dropped, so the implementation you have shown seems sufficient here.

I've simplified the drop implementation with another commit.